substrate/frame/revive/src/precompiles/builtin/xcm.rs

```diff
@@ -0,0 +1,169 @@
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Origin;
+use crate::{
+	precompiles::{AddressMatcher, Error, Ext, ExtWithInfo, Precompile},
+	Config,
+};
+use alloc::vec::Vec;
+use alloy_core::sol;
+use codec::{Encode, DecodeAll};
+use core::{marker::PhantomData, num::NonZero};
+use log::error;
+use sp_runtime::Weight;
+use xcm_builder::{ExecuteController, SendController};
+use xcm_executor::traits::WeightBounds;
+use alloy_core::sol_types::SolValue;
+
+pub use IXcm::IXcmCalls;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+sol! {
+	/// @title Defines all functions that can be used to interact with XCM
+	/// @author Tiago Bandeira
+	/// @dev Parameters MUST use SCALE codec serialisation
+	interface IXcm {
+		struct Weight {
+			uint64 refTime;
+			uint64 proofSize;
+		}
+
+		/// @notice Execute an XCM message locally with the caller's origin
+		/// @param message The XCM message to send
+		/// @param weight The maximum amount of weight to be used to execute the message
+		function xcmExecute(bytes calldata message, Weight calldata weight) external;
+
+		/// @notice Send an XCM message to a destination chain
+		/// @param destination The destination location, encoded according to the XCM format
+		/// @param message The XCM message to send
+		function xcmSend(bytes calldata destination, bytes calldata message) external;
+
+		/// @notice Given a message estimate the weight cost
+		/// @param message The XCM message to send
+		/// @returns weight estimated for sending the message
+		function weightMessage(bytes calldata message) external view returns(Weight weight);
+	}
+}
+
+impl<T: Config> Precompile for XcmPrecompile<T> {
+	type T = T;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) => {
+				frame_system::RawOrigin::Signed(account_id.clone()).into()
+			},
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let final_destination = xcm::VersionedLocation::decode_all(&mut &destination[..])
```

Comment:
This precompile will be updated to only support the latest versions, since `Versioned*` types only support the latest 3 XCM versions. With this, XCMs crafted in SCs should always be upgradable. Probably the pattern here (for mad people) is to put XCMs in a storage item so you can update them when needed. We need another higher level precompile to make everyone's lives easier